File: alphaforge/model/identifiers.py
# AlphaForge Identifier Types
"""
Identifier types for AlphaForge trading system.
"""

from typing import Optional
from alphaforge.core.exceptions import ValidationError


class InstrumentId:
    """Instrument identifier with symbol and venue."""
    
    def __init__(self, identifier: str):
        """
        Create instrument ID from symbol.venue format.
        
        Args:
            identifier: String in format "SYMBOL.VENUE"
        """
        if not identifier or '.' not in identifier:
            raise ValidationError(f"Invalid instrument identifier format: {identifier}")
            
        parts = identifier.split('.')
        if len(parts) != 2:
            raise ValidationError(f"Invalid instrument identifier format: {identifier}")
            
        self._symbol = parts[0].upper()
        self._venue = parts[1].upper()
        self._value = f"{self._symbol}.{self._venue}"
        
        if not self._symbol or not self._venue:
            raise ValidationError(f"Empty symbol or venue in identifier: {identifier}")
    
    @property
    def symbol(self) -> str:
        """Get the symbol component."""
        return self._symbol
    
    @property 
    def venue(self) -> str:
        """Get the venue component."""
        return self._venue
    
    @property
    def value(self) -> str:
        """Get the full identifier string."""
        return self._value
    
    @classmethod
    def from_parts(cls, symbol: str, venue: str) -> "InstrumentId":
        """Create from separate symbol and venue."""
        return cls(f"{symbol}.{venue}")
    
    def __str__(self) -> str:
        return self._value
    
    def __repr__(self) -> str:
        return f"InstrumentId('{self._value}')"
    
    def __eq__(self, other) -> bool:
        if not isinstance(other, InstrumentId):
            return False
        return self._value == other._value
    
    def __hash__(self) -> int:
        return hash(self._value)


class AccountId:
    """Account identifier with issuer and number."""
    
    def __init__(self, issuer: str, number: str):
        """
        Create account ID.
        
        Args:
            issuer: Account issuer (exchange/broker)
            number: Account number
        """
        if not issuer or not number:
            raise ValidationError("Issuer and number cannot be empty")
            
        self._issuer = issuer.upper()
        self._number = number
        self._value = f"{self._issuer}-{self._number}"
    
    @property
    def issuer(self) -> str:
        """Get the issuer component."""
        return self._issuer
    
    @property
    def number(self) -> str:
        """Get the account number."""
        return self._number
    
    @property
    def value(self) -> str:
        """Get the full identifier."""
        return self._value
    
    def __str__(self) -> str:
        return self._value
    
    def __repr__(self) -> str:
        return f"AccountId('{self._value}')"
    
    def __eq__(self, other) -> bool:
        if not isinstance(other, AccountId):
            return False
        return self._value == other._value
    
    def __hash__(self) -> int:
        return hash(self._value)


class ClientOrderId:
    """Client order identifier."""
    
    def __init__(self, value: str):
        """
        Create client order ID.
        
        Args:
            value: Order identifier string
        """
        if not value:
            raise ValidationError("Client order ID cannot be empty")
        if len(value) > 64:
            raise ValidationError("Client order ID too long (max 64 characters)")
            
        self._value = value
    
    @property
    def value(self) -> str:
        """Get the identifier value."""
        return self._value
    
    @classmethod
    def generate(cls) -> "ClientOrderId":
        """Generate a new UUID-based client order ID."""
        from alphaforge.core import uuid4_new
        return cls(uuid4_new())
    
    def __str__(self) -> str:
        return self._value
    
    def __repr__(self) -> str:
        return f"ClientOrderId('{self._value}')"
    
    def __eq__(self, other) -> bool:
        if not isinstance(other, ClientOrderId):
            return False
        return self._value == other._value
    
    def __hash__(self) -> int:
        return hash(self._value)


class VenueOrderId:
    """Venue order identifier."""
    
    def __init__(self, value: str):
        """Create venue order ID."""
        if not value:
            raise ValidationError("Venue order ID cannot be empty")
        self._value = value
    
    @property
    def value(self) -> str:
        """Get the identifier value.""" 
        return self._value
    
    def __str__(self) -> str:
        return self._value
    
    def __repr__(self) -> str:
        return f"VenueOrderId('{self._value}')"
    
    def __eq__(self, other) -> bool:
        if not isinstance(other, VenueOrderId):
            return False
        return self._value == other._value
    
    def __hash__(self) -> int:
        return hash(self._value)


class TradeId:
    """Trade identifier."""
    
    def __init__(self, value: str):
        """Create trade ID."""
        if not value:
            raise ValidationError("Trade ID cannot be empty")
        self._value = value
    
    @property
    def value(self) -> str:
        """Get the identifier value."""
        return self._value
    
    def __str__(self) -> str:
        return self._value
    
    def __repr__(self) -> str:
        return f"TradeId('{self._value}')"
    
    def __eq__(self, other) -> bool:
        if not isinstance(other, TradeId):
            return False
        return self._value == other._value
    
    def __hash__(self) -> int:
        return hash(self._value)


class PositionId:
    """Position identifier."""
    
    def __init__(self, value: str):
        """Create position ID."""
        if not value:
            raise ValidationError("Position ID cannot be empty")
        self._value = value
    
    @property
    def value(self) -> str:
        """Get the identifier value."""
        return self._value
    
    def __str__(self) -> str:
        return self._value
    
    def __repr__(self) -> str:
        return f"PositionId('{self._value}')"
    
    def __eq__(self, other) -> bool:
        if not isinstance(other, PositionId):
            return False
        return self._value == other._value
    
    def __hash__(self) -> int:
        return hash(self._value)


class StrategyId:
    """Strategy identifier."""
    
    def __init__(self, value: str):
        """Create strategy ID."""
        if not value:
            raise ValidationError("Strategy ID cannot be empty")
        if len(value) > 64:
            raise ValidationError("Strategy ID too long (max 64 characters)")
        self._value = value
    
    @property
    def value(self) -> str:
        """Get the identifier value."""
        return self._value
    
    def __str__(self) -> str:
        return self._value
    
    def __repr__(self) -> str:
        return f"StrategyId('{self._value}')"
    
    def __eq__(self, other) -> bool:
        if not isinstance(other, StrategyId):
            return False
        return self._value == other._value
    
    def __hash__(self) -> int:
        return hash(self._value)


class TraderId:
    """Trader identifier."""
    
    def __init__(self, value: str):
        """Create trader ID.""" 
        if not value:
            raise ValidationError("Trader ID cannot be empty")
        if len(value) > 64:
            raise ValidationError("Trader ID too long (max 64 characters)")
        self._value = value
    
    @property
    def value(self) -> str:
        """Get the identifier value."""
        return self._value
    
    def __str__(self) -> str:
        return self._value
    
    def __repr__(self) -> str:
        return f"TraderId('{self._value}')"
    
    def __eq__(self, other) -> bool:
        if not isinstance(other, TraderId):
            return False
        return self._value == other._value
    
    def __hash__(self) -> int:
        return hash(self._value)

File: crates/model/src/identifiers.rs
//! Identifier types for AlphaForge trading system

use std::fmt;
use serde::{Serialize, Deserialize};
use alphaforge_core::uuid::UUID4;

/// Instrument identifier
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct InstrumentId {
    symbol: String,
    venue: String,
    raw: String, // Cached full identifier
}

impl InstrumentId {
    /// Create a new instrument ID from symbol.venue format
    pub fn new(identifier: &str) -> Result<Self, IdentifierError> {
        let parts: Vec<&str> = identifier.split('.').collect();
        if parts.len() != 2 {
            return Err(IdentifierError::InvalidFormat(identifier.to_string()));
        }
        
        let symbol = parts[0].to_uppercase();
        let venue = parts[1].to_uppercase();
        
        if symbol.is_empty() || venue.is_empty() {
            return Err(IdentifierError::EmptyComponent(identifier.to_string()));
        }
        
        let raw = format!("{}.{}", symbol, venue);
        
        Ok(Self {
            symbol,
            venue,
            raw,
        })
    }
    
    /// Get the symbol component
    pub fn symbol(&self) -> &str {
        &self.symbol
    }
    
    /// Get the venue component
    pub fn venue(&self) -> &str {
        &self.venue
    }
    
    /// Get the full identifier string
    pub fn value(&self) -> &str {
        &self.raw
    }
    
    /// Create from separate symbol and venue
    pub fn from_parts(symbol: &str, venue: &str) -> Result<Self, IdentifierError> {
        Self::new(&format!("{}.{}", symbol, venue))
    }
}

impl fmt::Display for InstrumentId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.raw)
    }
}

/// Account identifier
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct AccountId {
    issuer: String,
    number: String,
    raw: String,
}

impl AccountId {
    /// Create a new account ID
    pub fn new(issuer: &str, number: &str) -> Result<Self, IdentifierError> {
        if issuer.is_empty() || number.is_empty() {
            return Err(IdentifierError::EmptyComponent(format!("{}-{}", issuer, number)));
        }
        
        let raw = format!("{}-{}", issuer.to_uppercase(), number);
        
        Ok(Self {
            issuer: issuer.to_uppercase(),
            number: number.to_string(),
            raw,
        })
    }
    
    /// Get the issuer component
    pub fn issuer(&self) -> &str {
        &self.issuer
    }
    
    /// Get the account number
    pub fn number(&self) -> &str {
        &self.number
    }
    
    /// Get the full identifier
    pub fn value(&self) -> &str {
        &self.raw
    }
}

impl fmt::Display for AccountId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.raw)
    }
}

/// Client order identifier
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct ClientOrderId {
    value: String,
}

impl ClientOrderId {
    /// Create a new client order ID
    pub fn new(value: &str) -> Result<Self, IdentifierError> {
        if value.is_empty() {
            return Err(IdentifierError::EmptyComponent(value.to_string()));
        }
        
        if value.len() > 64 {
            return Err(IdentifierError::TooLong(value.to_string()));
        }
        
        Ok(Self {
            value: value.to_string(),
        })
    }
    
    /// Generate a new UUID-based client order ID
    pub fn generate() -> Self {
        Self {
            value: UUID4::new().to_string(),
        }
    }
    
    /// Get the identifier value
    pub fn value(&self) -> &str {
        &self.value
    }
}

impl fmt::Display for ClientOrderId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.value)
    }
}

/// Venue order identifier
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct VenueOrderId {
    value: String,
}

impl VenueOrderId {
    /// Create a new venue order ID
    pub fn new(value: &str) -> Result<Self, IdentifierError> {
        if value.is_empty() {
            return Err(IdentifierError::EmptyComponent(value.to_string()));
        }
        
        Ok(Self {
            value: value.to_string(),
        })
    }
    
    /// Get the identifier value
    pub fn value(&self) -> &str {
        &self.value
    }
}

impl fmt::Display for VenueOrderId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.value)
    }
}

/// Trade identifier
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct TradeId {
    value: String,
}

impl TradeId {
    /// Create a new trade ID
    pub fn new(value: &str) -> Result<Self, IdentifierError> {
        if value.is_empty() {
            return Err(IdentifierError::EmptyComponent(value.to_string()));
        }
        
        Ok(Self {
            value: value.to_string(),
        })
    }
    
    /// Get the identifier value
    pub fn value(&self) -> &str {
        &self.value
    }
}

impl fmt::Display for TradeId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.value)
    }
}

/// Position identifier
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct PositionId {
    value: String,
}

impl PositionId {
    /// Create a new position ID
    pub fn new(value: &str) -> Result<Self, IdentifierError> {
        if value.is_empty() {
            return Err(IdentifierError::EmptyComponent(value.to_string()));
        }
        
        Ok(Self {
            value: value.to_string(),
        })
    }
    
    /// Get the identifier value
    pub fn value(&self) -> &str {
        &self.value
    }
}

impl fmt::Display for PositionId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.value)
    }
}

/// Strategy identifier
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct StrategyId {
    value: String,
}

impl StrategyId {
    /// Create a new strategy ID
    pub fn new(value: &str) -> Result<Self, IdentifierError> {
        if value.is_empty() {
            return Err(IdentifierError::EmptyComponent(value.to_string()));
        }
        
        if value.len() > 64 {
            return Err(IdentifierError::TooLong(value.to_string()));
        }
        
        Ok(Self {
            value: value.to_string(),
        })
    }
    
    /// Get the identifier value
    pub fn value(&self) -> &str {
        &self.value
    }
}

impl fmt::Display for StrategyId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.value)
    }
}

/// Trader identifier
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct TraderId {
    value: String,
}

impl TraderId {
    /// Create a new trader ID
    pub fn new(value: &str) -> Result<Self, IdentifierError> {
        if value.is_empty() {
            return Err(IdentifierError::EmptyComponent(value.to_string()));
        }
        
        if value.len() > 64 {
            return Err(IdentifierError::TooLong(value.to_string()));
        }
        
        Ok(Self {
            value: value.to_string(),
        })
    }
    
    /// Get the identifier value
    pub fn value(&self) -> &str {
        &self.value
    }
}

impl fmt::Display for TraderId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.value)
    }
}

/// Identifier error types
#[derive(Debug, thiserror::Error)]
pub enum IdentifierError {
    #[error("Invalid format: {0}")]
    InvalidFormat(String),
    #[error("Empty component: {0}")]
    EmptyComponent(String),
    #[error("Too long: {0}")]
    TooLong(String),
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_instrument_id_creation() {
        let id = InstrumentId::new("BTCUSD.BINANCE").unwrap();
        assert_eq!(id.symbol(), "BTCUSD");
        assert_eq!(id.venue(), "BINANCE");
        assert_eq!(id.value(), "BTCUSD.BINANCE");
        
        // Test case normalization
        let id2 = InstrumentId::new("btcusd.binance").unwrap();
        assert_eq!(id2.value(), "BTCUSD.BINANCE");
    }
    
    #[test]
    fn test_instrument_id_validation() {
        assert!(InstrumentId::new("INVALID").is_err());
        assert!(InstrumentId::new(".BINANCE").is_err());
        assert!(InstrumentId::new("BTCUSD.").is_err());
        assert!(InstrumentId::new("").is_err());
    }
    
    #[test]
    fn test_account_id_creation() {
        let id = AccountId::new("BINANCE", "123456").unwrap();
        assert_eq!(id.issuer(), "BINANCE");
        assert_eq!(id.number(), "123456");
        assert_eq!(id.value(), "BINANCE-123456");
    }
    
    #[test]
    fn test_client_order_id_generation() {
        let id1 = ClientOrderId::generate();
        let id2 = ClientOrderId::generate();
        
        assert_ne!(id1.value(), id2.value());
        assert_eq!(id1.value().len(), 36); // UUID format
    }
}
